### PR TITLE
fix: event parsing iterator

### PR DESCRIPTION
**Pull request by yenfryherrerafeliz**
The EventParsingIterator implementation does not parse correctly events when their payload do not have the eventpayload property set. The current behavior is that if said property "eventpayload" is not present then it fallback to extracting the value from the headers, but this is not a reliable method for parsing event streams since seems that not all services uses this pattern to provide the events. One example is the InvokeModelWithResponseStream operation from bedrock-runtime service. To fix this we check if the event key is set in the headers, and if so then we pick the value from the headers, otherwise we set the event key value based on the event payload.

*Issues #2817, #2787*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by YujiroTakahashi** [comment]
I tested the revised content with the following code. 
It waits for completion just like the invokeModel() function.  
Would asynchronous processing be required separately?

```php
<?php

include '/opt/vendor/autoload.php';

use Aws\BedrockRuntime\BedrockRuntimeClient;

$apiClient = new BedrockRuntimeClient([
    'region'  => 'us-east-1',
]);
$payload = [
    'modelId' => 'anthropic.claude-v2',
    'contentType' => 'application/json',
    'accept' => '*/*',
    'body' => json_encode([
        'prompt'=> "\n\nHuman: Please give me a PHP sample using BedrockRuntimeClient. \n\nAssistant:",
        'max_tokens_to_sample'=> 1024,
        'temperature'=> 1.0,
        'top_k'=> 100,
        'top_p'=> 0.9,
        'stop_sequences'=> [
          "\n\nHuman:"
        ],
        'anthropic_version'=> 'bedrock-2023-05-31',
    ]),
];

$result = $apiClient->invokeModelWithResponseStream($payload);
$iterator = $result->get('body');
foreach ($iterator as $event) {
    if (isset($event['chunk'])) {
        $chunk = json_decode($event['chunk']['bytes']->getContents(), true);
        $data = json_decode(base64_decode($chunk['bytes']), true);
        ob_start();
        print_r($data);
        ob_end_flush();
        flush();
    }
}

```

```php
<?php

include '/opt/vendor/autoload.php';

use Aws\BedrockRuntime\BedrockRuntimeClient;

$apiClient = new BedrockRuntimeClient([
    'region'  => 'us-east-1',
]);
$payload = [
    'modelId' => 'anthropic.claude-v2',
    'contentType' => 'application/json',
    'accept' => '*/*',
    'body' => json_encode([
        'prompt'=> "\n\nHuman: Please give me a PHP sample using BedrockRuntimeClient. \n\nAssistant:",
        'max_tokens_to_sample'=> 1024,
        'temperature'=> 1.0,
        'top_k'=> 100,
        'top_p'=> 0.9,
        'stop_sequences'=> [
          "\n\nHuman:"
        ],
        'anthropic_version'=> 'bedrock-2023-05-31',
    ]),
];
$promise = $apiClient->invokeModelWithResponseStreamAsync($payload)->then(
    function ($result) {
        $iterator = $result->get('body');
        foreach ($iterator as $event) {
            if (isset($event['chunk'])) {
                $chunk = json_decode($event['chunk']['bytes']->getContents(), true);
                $data = json_decode(base64_decode($chunk['bytes']), true);
                ob_start();
                print_r($data);
                ob_end_flush();
                flush();
            }
        }
    },
    function ($reason) {
        echo "Failed: " . $reason;
    }
);
$promise->wait();
```

```
$ php stream.php
Array
(
    [completion] =>  Here
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  is a simple PHP code snippet using the BedrockRuntimeClient
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  to invoke a bedrock function:

` ` ` php
<?
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => php

use Bedrock\Runtime\
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => BedrockRuntimeClient;

$client = new Bedrock
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => RuntimeClient('https://bedrock.example.com'); 

$
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => request = [
    'moduleName
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => ' => 'myModule',
    'methodName' => 'helloWorld
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => ',
    'params'
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  => ['name' => 'John']
];

$response = $client->invoke
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => ($request);

print_r($response);
` ` `

The
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  key steps are:

1. Create a BedrockRuntime
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] => Client instance, passing the Bedrock
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  URL.

2. Create the request
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  data with the module name, method name and
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  parameters. 

3. Call the invoke() method on the client, passing
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  the request.

4. The response contains the result of calling the
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  bedrock function.

Let me know if you need any
    [stop_reason] => 
    [stop] => 
)
Array
(
    [completion] =>  clarification or have additional questions!
    [stop_reason] => stop_sequence
    [stop] => 

Human:
)
```

## SDK version used
3.287.0

## Environment details
```shell
$ php -v
PHP 8.1.2-1ubuntu2.14 (cli) (built: Aug 18 2023 11:41:11) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.2, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.2-1ubuntu2.14, Copyright (c), by Zend Technologies
```

File: src/Api/Parser/EventParsingIterator.php [hunk 1]
```diff
@@ -69,36 +69,43 @@ private function parseEvent(array $event)
...
                 return $this->parseError($event);
             }
+
             if ($event['headers'][':message-type'] !== 'event') {
                 throw new ParserException('Failed to parse unknown message type.');
             }
         }
 
-        if (empty($event['headers'][':event-type'])) {
+        $eventType = $event['headers'][':event-type'] ?? null;
```
Comment (stobrien89) [inline src/Api/Parser/EventParsingIterator.php:78]: Are we guaranteed this will always be set?

File: src/Api/Parser/EventParsingIterator.php [hunk 2]
```diff
@@ -69,36 +69,43 @@ private function parseEvent(array $event)
...
-            if (!empty($details['eventpayload'])) {
-                $payloadShape = $eventShape->getMember($shape);
-                if ($payloadShape['type'] === 'blob') {
-                    $parsedEvent[$shape] = $event['payload'];
+        foreach ($eventShape['members'] as $eventShapeMemberKey => $eventShapeMemberProps) {
+            $memberShape = $eventShape->getMember($eventShapeMemberKey);
+            if (!empty($eventShapeMemberProps['eventheader'])) {
+                $parsedEvent[$eventShapeMemberKey] = $event['headers'][$eventShapeMemberKey];
+            } elseif (!empty($eventShapeMemberProps['eventpayload'])) {
+                if ($memberShape['type'] === 'blob') {
```
Comment (stobrien89) [inline src/Api/Parser/EventParsingIterator.php:91]: pretty sure you can use a `getType` call here

File: src/Api/Parser/EventParsingIterator.php [hunk 3]
```diff
@@ -69,36 +69,43 @@ private function parseEvent(array $event)
...
 
+        $eventShape = $this->shape->getMember($eventType);
+        $eventPayload = $event['payload'];
         $parsedEvent = [];
-        foreach ($eventShape['members'] as $shape => $details) {
-            if (!empty($details['eventpayload'])) {
-                $payloadShape = $eventShape->getMember($shape);
-                if ($payloadShape['type'] === 'blob') {
-                    $parsedEvent[$shape] = $event['payload'];
+        foreach ($eventShape['members'] as $eventShapeMemberKey => $eventShapeMemberProps) {
```
Comment (stobrien89) [inline src/Api/Parser/EventParsingIterator.php:86]: I'd check to see if you can use a getMembers call here.  It might not be available to these shapes.

**Review by stobrien89 — CHANGES_REQUESTED** [review]
Looking good so far, just a few suggestions.  Let's get the tests ready and go from there.

File: src/Api/Parser/EventParsingIterator.php (hunk 1 above)
Comment (yenfryherrerafeliz) [reply src/Api/Parser/EventParsingIterator.php]: Actually my evaluation is wrong. We should even check for when this property is set but empty, which also should fail in that scenario.

File: src/Api/Parser/EventParsingIterator.php (hunk 3 above)
Comment (yenfryherrerafeliz) [reply src/Api/Parser/EventParsingIterator.php]: Done!

File: src/Api/Parser/EventParsingIterator.php (hunk 2 above)
Comment (yenfryherrerafeliz) [reply src/Api/Parser/EventParsingIterator.php]: Done!

**Review by stobrien89 — CHANGES_REQUESTED** [review]
Requested a few changes

**Review by stobrien89 — CHANGES_REQUESTED** [review]

File: src/Api/Parser/EventParsingIterator.php [hunk 4]
```diff
@@ -69,37 +69,81 @@ private function parseEvent(array $event)
...
+                    $parsedPayload[$memberName] = $this->parser->parseMemberFromStream(
+                        $payload,
+                        $memberShape,
                         null
                     );
                 }
-            } else {
-                $parsedEvent[$shape] = $event['headers'][$shape];
+
+                break;
```
[the remark this answers is not in the document]
Comment (stobrien89) [reply src/Api/Parser/EventParsingIterator.php:134]: Nevermind:

>If the eventPayload trait is applied to a structure member, then all other members of the structure MUST be marked with the eventHeader trait.

**Review by stobrien89 — APPROVED** [review]
Just needs a changelog entry 🚢

**Pull request merged**